### PR TITLE
Diffuser change upgraded to 0.26.3 along with MLPERF SD XL support

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
1. MLPERF SD XL inference related change unstreamed
    2. All Diffuser changes upgraded to 0.26.3
JH...

### CODE REVIEW

**Comment by sywangyi** [comment]
@yuanwu2017 @cfgfung please be ware of it.

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:611]: Why hard code 18, do we assume 20 timesteps/inference steps?

File: optimum/habana/diffusers/models/attention_processor.py
Comment (sywangyi) [inline optimum/habana/diffusers/models/attention_processor.py:42]: why is softmax_fp8 used here? support stable diffuser for fp8?

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (sywangyi) [reply optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:130]: StableDiffusionXLPipeline_HPU is not used in any example.

File: optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py
Comment (sywangyi) [inline optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py:239]: what's is this hpu_opt for? see **self.hpu_opt = False** in __init__, not found "self.hpu_opt = True" in any place.

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (ANSHUMAN87) [reply optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:130]: StableDiffusionXLPipeline_HPU implementation is used by MLPERF SD XL workload published in model garden. So the usage in optimum-habana wont be visible. 

JH...!

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (ANSHUMAN87) [reply optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:611]: Yes there will be only 20 steps/inference in case of MLPERF SD XL .

JH...!

File: optimum/habana/diffusers/models/attention_processor.py
Comment (ANSHUMAN87) [reply optimum/habana/diffusers/models/attention_processor.py:42]: Current support is for FP8 only.

File: optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py
Comment (ANSHUMAN87) [reply optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py:239]: "self.hpu_opt = True" will be done in MLPERF SD XL repo where we instantiate the class. In optimum-habana, wont be visible. By default is set to Flase, so that current flow is intact in optimum-habana.

JH...!

Comment (libinta) [reply optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py:239]: @ANSHUMAN87 do you mean optimum-habana doesn't need hpu_opt =true and no need for optimum-habana ? or you don't have method to set it to true for optimum-habana?

File: optimum/habana/diffusers/models/attention_processor.py
Comment (libinta) [inline optimum/habana/diffusers/models/attention_processor.py:58]: @ANSHUMAN87 the scoped optimization for matmu;/softmax is for fp8? and how to invoke it for bf16 model run?

File: optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py
Comment (ANSHUMAN87) [reply optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py:239]: As of now, optimum-habana no need to set to True, in case in future needed, user can instantiate the class and set to True, if user wants to use those optimizations. Those optimizations are applicable only for MLPERF SD-XL repo currently.

File: optimum/habana/diffusers/models/attention_processor.py
Comment (kumarans-ai) [reply optimum/habana/diffusers/models/attention_processor.py:58]: @libinta , we have both options when you pass scale as None automatically kernel will operate in bf16. Please refer softmax API for further

File: optimum/habana/diffusers/models/unet_2d_condition.py
Comment (dsocek) [inline optimum/habana/diffusers/models/unet_2d_condition.py:14]: Can you, please, fix indentation?   (remove 4 spaces from all lines in this function)

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1191]: Can you put back the original name, and rename the `retrieve_timesteps` specific to ML perf _hpu class instead?

**Review by dsocek — APPROVED** [review]
LGTM as long as OK with CODEOWNERS to add an extra MLPerf-specific class to OH

**Comment by libinta** [comment]
@ANSHUMAN87  @dsocek what are the change needed to make sure we can utilize the mlperf optimization?

**Comment by yeonsily** [comment]
@ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

**Comment by dsocek** [comment]
> @ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

@libinta, @yeonsily I think that @ANSHUMAN87 is better to answer this as he created the code/PR.  We can consider if some of his optimization strategies can be applied to a normal OH Stable Diffusion pipeline (but this means no hard coded params in pipeline as users may select/run different use-cases)

**Comment by ANSHUMAN87** [comment]
> @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?

We have MLPERF specific optimizations only in 2 places, other places are combined for both:
1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**

Please let me know, if need any more info.

JH...!

**Comment by ANSHUMAN87** [comment]
> @ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

@yeonsily I had run ruff in my local, all errors are handled already. But the import error reported in this PR, I am unable to reproduce in my local. Any suggestions ?

JH...!

**Comment by dsocek** [comment]
> > @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?
> 
> We have MLPERF specific optimizations only in 2 places, other places are combined for both:
> 
> 1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
> 2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**
> 
> Please let me know, if need any more info.
> 
> JH...!

@ANSHUMAN87 Thanks, but more info is needed here.  Can you get into the details of the strategies in **StableDiffusionXLPipeline_HPU** class, and what is different when compared to general class **GaudiStableDiffusionXLPipeline**.

**Comment by ANSHUMAN87** [comment]
> > > @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?
> > 
> > 
> > We have MLPERF specific optimizations only in 2 places, other places are combined for both:
> > 
> > 1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
> > 2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**
> > 
> > Please let me know, if need any more info.
> > JH...!
> 
> @ANSHUMAN87 Thanks, but more info is needed here. Can you get into the details of the strategies in **StableDiffusionXLPipeline_HPU** class, and what is different when compared to general class **GaudiStableDiffusionXLPipeline**.

Below are the high level changes done in STableDiffusionXLPipeline_HPU:
   - Added mark_step after unet Model in the pipeline
   - Added attention_processor implementation
   - Added unet forward implementation
   - Added quantization support

JH...!

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (regisss) [reply optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:135]: Let's move this new class and all the related changes in this file to a new file called `pipeline_stable_diffusion_xl_mlperf.py` please. It will be clearer and easier to maintain.

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:52]:
```suggestion
from ...diffusers import GaudiDiffusionPipeline
from ...diffusers.models.attention_processor import (
    AttentionProcessor,
    AttnProcessor2_0,
    ScaledDotProductAttention,
)
from ...models.unet_2d_condition import gaudi_unet_2d_condition_model_forward
from ..pipelines.stable_diffusion.pipeline_stable_diffusion import (
    retrieve_timesteps,
)
from ....transformers.gaudi_configuration import GaudiConfig
from ....utils import HabanaProfile, speed_metrics, warmup_inference_steps_time_adjustment
```
Please use relative imports in the source code

**Review by regisss — COMMENTED** [review]
I left a few comments, the main one being about moving all the code added to `pipeline_stable_diffusion_xl.py` to a new file so that the codebase is clearer and easier to maintain.

**Comment by ANSHUMAN87** [comment]
@regisss your comments are addressed now. Thanks!

JH...!

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py:630]: Where is `quantized` defined for the class?

File: optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py:187]: Why are we not inheriting gaudi-specific class members/configs from `GaudiDiffusionPipeline`?

Comment (ANSHUMAN87) [reply optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py:187]: @dsocek I did not find any specific need to use GaudiDiffusionPipeline. But I am open for suggestions.

**Comment by imangohari1** [comment]
@ANSHUMAN87 
Hi, 
Could you rebase and re-test this with the main? The main diffuser is now that 0.29.2 and some changes are needed for this PR. Thanks.

**Comment by regisss** [comment]
Closing as these changes were added to https://github.com/huggingface/optimum-habana/pull/1204.

**Pull request closed without merging**
